libethereum/Block.cpp

```diff
@@ -136,47 +136,59 @@ void Block::noteChain(BlockChain const& _bc)
     }
 }
 
-PopulationStatistics Block::populateFromChain(BlockChain const& _bc, h256 const& _h, ImportRequirements::value _ir)
+void Block::populateFromChain(BlockChain const& _bc, h256 const& _h)
 {
-    noteChain(_bc);
-
-    PopulationStatistics ret { 0.0, 0.0 };
-
     if (!_bc.isKnown(_h))
     {
-        // Might be worth throwing here.
-        cwarn << "Invalid block given for state population: " << _h;
+        cerr << "Invalid block given for state population: " << _h;
         BOOST_THROW_EXCEPTION(BlockNotFound() << errinfo_target(_h));
     }
 
-    auto b = _bc.block(_h);
-    BlockHeader bi(b);		// No need to check - it's already in the DB.
-    if (bi.number())
+    auto const& blockBytes = _bc.block(_h);
+    m_currentBytes = blockBytes;
+
+    // Set block headers
+    auto const& blockHeader = BlockHeader{blockBytes};
+    m_currentBlock = blockHeader;
+
+    if (blockHeader.number())
+        m_previousBlock = _bc.info(blockHeader.parentHash());
+    else
+        m_previousBlock = m_currentBlock;
+
+    // Set state root and precommit state
+    //
+    // First check for database corruption by looking up the state root in the state database. Note
+    // that we don't technically need to do this since if the state DB is corrupt setting a new
+    // state root will throw anyway, but checking here enables us to log a user-friendly error
+    // message.
+    if (m_state.db().lookup(blockHeader.stateRoot()).empty())
     {
-        // Non-genesis:
-
-        // 1. Start at parent's end state (state root).
-        BlockHeader bip(_bc.block(bi.parentHash()));
-        sync(_bc, bi.parentHash(), bip);
-
-        // 2. Enact the block's transactions onto this state.
-        m_author = bi.author();
-        Timer t;
-        auto vb = _bc.verifyBlock(&b, function<void(Exception&)>(), _ir | ImportRequirements::TransactionBasic);
-        ret.verify = t.elapsed();
-        t.restart();
-        enact(vb, _bc);
-        ret.enact = t.elapsed();
+        cerr << "Unable to populate block " << blockHeader.hash() << " - state root "
+             << blockHeader.stateRoot() << " not found in database.";
+        cerr << "Database corrupt: contains block without state root: " << blockHeader;
+        cerr << "Try rescuing the database by running: eth --rescue";
+        BOOST_THROW_EXCEPTION(InvalidStateRoot() << errinfo_target(blockHeader.stateRoot()));
     }
-    else
+
+    m_state.setRoot(blockHeader.stateRoot());
+    m_precommit = m_state;
```

Comment:
Was it setting `m_precommit` before?

Reply:
Yes - see line 115: https://github.com/ethereum/aleth/blob/9f31c8b1f720a33cd1e8f7467b780464f0802af6/libethereum/Block.cpp#L101-L120